File: frontend/src/types/MetricType.ts
export type MetricType =
    "cpu_usage"
    | "used_memory_mb"
    | "used_swap_mb"
    | "drive_used_space"
    | "network_upload"
    | "network_download"


export function prettyMetricType(value: MetricType): string {
    switch (value) {
        case "cpu_usage":
            return "CPU Usage (%)"
        case "used_memory_mb":
            return "Used Memory (MB)"
        case "used_swap_mb":
            return "Used Swap (MB)"
        case "drive_used_space":
            return "Drive Used Space (GB)"
        case "network_upload":
            return "Network Upload (kB)"
        case "network_download":
            return "Network Download (kB)"
    }
}

export function metricTypeUnit(value: MetricType): string {
    switch (value) {
        case "cpu_usage":
            return "%"
        case "used_memory_mb":
        case "used_swap_mb":
            return "MB"
        case "drive_used_space":
            return "GB"
        case "network_upload":
        case "network_download":
            return "kB"
    }

}

File: backend/crates/shared/src/enums/metric_type.rs
use std::{fmt::Display, str::FromStr};

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub enum MetricType {
    CpuUsage,
    UsedMemoryMb,
    UsedSwapMb,
    DriveUsedSpace,
    NetworkDownload,
    NetworkUpload,
}

impl MetricType {
    pub fn to_string_pretty(&self) -> Option<String> {
        let val = match self {
            Self::CpuUsage => "CPU Usage (%)",
            Self::UsedMemoryMb => "Used Memory (MB)",
            Self::UsedSwapMb => "Used Swap (MB)",
            Self::DriveUsedSpace => "Used Drive Space (GB)",
            Self::NetworkDownload => "Network Download (KB)",
            Self::NetworkUpload => "Network Upload (KB)",
        };

        String::from_str(val).ok()
    }

    pub fn unit(&self) -> Option<String> {
        let val = match &self {
            MetricType::CpuUsage => "%",
            MetricType::UsedMemoryMb | MetricType::UsedSwapMb => "MB",
            MetricType::DriveUsedSpace => "GB",
            MetricType::NetworkDownload | MetricType::NetworkUpload => "kB",
        };

        String::from_str(val).ok()
    }
}

impl FromStr for MetricType {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "cpu_usage" => Ok(Self::CpuUsage),
            "used_memory_mb" => Ok(Self::UsedMemoryMb),
            "used_swap_mb" => Ok(Self::UsedSwapMb),
            "drive_used_space" => Ok(Self::DriveUsedSpace),
            "network_upload" => Ok(Self::NetworkUpload),
            "network_download" => Ok(Self::NetworkDownload),
            _ => Err(s.to_string()),
        }
    }
}

impl Display for MetricType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let value = match self {
            MetricType::CpuUsage => "cpu_usage",
            MetricType::UsedMemoryMb => "used_memory_mb",
            MetricType::UsedSwapMb => "used_swap_mb",
            MetricType::DriveUsedSpace => "drive_used_space",
            MetricType::NetworkDownload => "network_download",
            MetricType::NetworkUpload => "network_upload",
        };

        write!(f, "{}", value)
    }
}
